lua/ulx/modules/urs_server.lua

```diff
@@ -1,193 +1,267 @@
+
+    -- true, false, or nil
+    return rawget( restrictionTypes, what )
+end
+local cachedCheck = URS.CachedCheck
+
+-- Caches the given check and returns the given result
+function URS.CacheCheck(ply, restrictionType, what, result)
+    ply.URS_CacheCheck = ply.URS_CacheCheck or {}
+
+    local existing = ply.URS_CacheCheck
+    local restrictionTypes = rawget( existing, restrictionType )
+
+    if not restrictionTypes then
+        rawset( existing, restrictionType, {} )
+        return result
+    end
+
+    rawset( restrictionTypes, what, result )
+
+    return result
+end
+local cacheCheck = URS.CacheCheck
+
+function URS.Check(ply, restrictionType, what)
+    local cachedResult = cachedCheck( ply, restrictionType, what )
+    if cachedResult ~= nil then
+        PrintRestricted( ply, restrictionType, what )
+        return cachedResult
+    end
+
+    what = stringLower(what)
+    local restrictionTypePlural = restrictionType .. "s"
+    local group = ply:GetUserGroup()
+
+    local restriction = false
+
+    local ursRestrictions = rawget( URS, "restrictions" )
+    local typeRestrictions = ursRestrictions and rawget( ursRestrictions, restrictionType )
+    local theseRestrictions = typeRestrictions and rawget( typeRestrictions, what )
+
+    if theseRestrictions then
+        restriction = theseRestrictions
+    end
+
+    if restriction then
+        local hasGroup = rawget( restriction, group )
+        local hasPlayer = rawget( restriction, ply:SteamID() )
+
+        if rawget( restriction, "*" ) then
+            if not ( hasGroup or hasPlayer ) then
+                PrintRestricted( ply, restrictionType, what )
+                return cacheCheck( ply, restrictionType, what, false )
+            end
+        elseif hasGroup or hasPlayer then
+            PrintRestricted( ply, restrictionType, what )
+            return cacheCheck( ply, restrictionType, what, false )
+        end
+    end
+
+    local allRestrictions = rawget( ursRestrictions, "all" )
+    local allTypeRestrictions = allRestrictions and rawget( allRestrictions, restrictionType )
+    local hasGroup = allTypeRestrictions and rawget( allTypeRestrictions, group )
+
+    if hasGroup then
+        tsayError(ply, "Your rank is restricted from all " .. restrictionTypePlural)
+        return cacheCheck( ply, restrictionType, what, false )
+    end
+
+    local ursTypes = rawget( URS, "types" )
+    local ursTypeLimits = rawget( ursTypes, "limitsMap" )
+    local limitsHasType = rawget( ursTypeLimits, restrictionType )
+    if not limitsHasType then return end
+
+    local ursLimits = rawget( URS, "limits" )
+    local typeLimits = rawget( ursLimits, restrictionType )
+    if not typeLimits then return end
+
+    local playerTypeLimit = rawget( typeLimits, ply:SteamID() )
+    local groupTypeLimit = rawget( typeLimits, group )
+    if not playerTypeLimit and not groupTypeLimit then return end
+
+    -- TODO: Should this be an elseif? Shouldn't both cases be possible?
```

Comment:
You can only return once, if both cases can exist playerlimit should go first. That's how regular URS does it too.